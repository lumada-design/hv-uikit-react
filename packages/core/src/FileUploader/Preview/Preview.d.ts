import { ButtonProps, StandardProps } from "@mui/material";

export type HvFileUploaderPreviewClassKey = "previewButton" | "overlay";

export interface PreviewProps extends StandardProps<ButtonProps, HvFileUploaderPreviewClassKey> {
  /**
   * Callback executed when the preview is unmounted.
   *
   * Should be used for cleaning up client-side image URLs created by `URL.createObjectURL()`.
   */
  onUnload?: void;
  /**
   * If `true`, doesn't show an overlay on top of the preview when hovering.
   *
   * Only applies when `onClick` is set.
   */
  disableOverlay?: boolean;
}

export default function HvFileUploaderPreview(props: PreviewProps): JSX.Element | null;
