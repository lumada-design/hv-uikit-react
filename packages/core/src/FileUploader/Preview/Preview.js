import React, { useEffect } from "react";
import PropTypes from "prop-types";

import { withStyles } from "@mui/styles";

import { Preview as PreviewIcon } from "@hitachivantara/uikit-react-icons";

import { HvButton } from "../..";

import styles from "./styles";

/**
 * The `HvFileUploaderPreview` component is available to facilitate the styling
 * of the button (when clickable) and the detection of image unloading.
 */
const Preview = (props) => {
  const { children, classes, onClick, onUnload, disableOverlay = false, ...others } = props;

  useEffect(() => {
    return () => {
      onUnload?.();
    };
  }, [onUnload]);

  if (onClick) {
    return (
      <HvButton icon className={classes.previewButton} onClick={onClick} {...others}>
        {children}
        {!disableOverlay && (
          <div className={classes.overlay} aria-hidden="true">
            <PreviewIcon />
          </div>
        )}
      </HvButton>
    );
  }

  return children;
};

Preview.propTypes = {
  /**
   * Content that represents the preview of an uploaded file.
   */
  children: PropTypes.node.isRequired,
  /**
   * Callback executed when the preview is clicked.
   *
   * When not null, the preview content will be enclosed in a button.
   *
   * @param event React.MouseEvent<HTMLButtonElement>
   */
  onClick: PropTypes.func,
  /**
   * Callback executed when the preview is unmounted.
   *
   * Should be used for cleaning up client-side image URLs created by `URL.createObjectURL()`.
   */
  onUnload: PropTypes.func,
  /**
   * If `true`, doesn't show an overlay on top of the preview when hovering.
   *
   * Only applies when `onClick` is set.
   */
  disableOverlay: PropTypes.bool,
  /**
   * A Jss Object used to override or extend the styles applied to the component.
   */
  classes: PropTypes.shape({
    /**
     * Styles to apply to the button when present.
     */
    previewButton: PropTypes.string,
    /**
     * Styles to apply to the overlay shown on hover.
     */
    overlay: PropTypes.string,
  }).isRequired,
};

export default withStyles(styles, { name: "HvFileUploaderPreview" })(Preview);
