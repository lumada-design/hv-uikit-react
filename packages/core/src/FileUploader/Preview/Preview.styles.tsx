import { theme } from "@hitachivantara/uikit-styles";

import { createClasses } from "../../utils/classes";

export const { staticClasses, useClasses } = createClasses(
  "HvFileUploaderPreview",
  {
    previewButton: {
      position: "relative",
      width: "48px",
      height: "48px",
      "&:hover *": {
        display: "flex",
      },
    },
    overlay: {
      position: "absolute",
      inset: 0,
      display: "none",
      justifyContent: "center",
      alignItems: "center",
      backgroundColor: theme.colors.primary_20,
      opacity: "1",
      borderRadius: theme.radii.base,
    },
  },
);
