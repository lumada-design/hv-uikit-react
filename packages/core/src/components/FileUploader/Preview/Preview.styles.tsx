import styled from "@emotion/styled";
import { Preview } from "@hitachivantara/uikit-react-icons";
import { theme } from "@hitachivantara/uikit-styles";
import { HvButton, HvButtonProps } from "@core/components";
import fileUploaderPreviewClasses from "./previewClasses";
import { forwardRef, Ref } from "react";

export const StyledOverlay = styled("div")({
  position: "absolute",
  top: 0,
  bottom: 0,
  left: 0,
  right: 0,
  display: "none",
  justifyContent: "center",
  alignItems: "center",
  backgroundColor: theme.fileUploader.preview.overlayColor,
  opacity: theme.fileUploader.preview.overlayOpacity,
  borderRadius: theme.fileUploader.preview.overlayBorderRadius,

  [`.${fileUploaderPreviewClasses.previewButton}:hover &`]: {
    display: "flex",
  },
});

export const StyledPreviewIcon = styled(Preview)({
  position: "absolute",
  left: "50%",
  top: "50%",
  transform: "translate(-50%, -50%)",
  display: "none",

  [`.${fileUploaderPreviewClasses.previewButton}:hover &`]: {
    display: "flex",
  },
});

export const StyledButton = styled(
  forwardRef((props: HvButtonProps, ref?: Ref<HTMLButtonElement>) => {
    return <HvButton {...props} ref={ref} />;
  })
)({
  position: "relative",
  width: theme.fileUploader.preview.buttonSize,
  height: theme.fileUploader.preview.buttonSize,
});
