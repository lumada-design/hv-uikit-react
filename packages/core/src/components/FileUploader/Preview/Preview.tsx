import { clsx } from "clsx";
import { HvButtonProps } from "@core/components";
import { useEffect } from "react";
import fileUploaderPreviewClasses, {
  HvFileUploaderPreviewClasses,
} from "./previewClasses";
import {
  StyledButton,
  StyledOverlay,
  StyledPreviewIcon,
} from "./Preview.styles";

export type HvFileUploaderPreviewProps = Omit<
  HvButtonProps,
  "children" | "classes"
> & {
  /**
   * Content that represents the preview of an uploaded file.
   */
  children: React.ReactElement;
  /**
   * Callback executed when the preview is unmounted.
   *
   * Should be used for cleaning up client-side image URLs created by `URL.createObjectURL()`.
   */
  onUnload?: () => void;
  /**
   * If `true`, doesn't show an overlay on top of the preview when hovering.
   *
   * Only applies when `onClick` is set.
   */
  disableOverlay?: boolean;
  /**
   * A Jss Object used to override or extend the styles applied to the component.
   */
  classes?: HvFileUploaderPreviewClasses;
};

/**
 * The `HvFileUploaderPreview` component is available to facilitate the styling
 * of the button (when clickable) and the detection of image unloading.
 */
export const HvFileUploaderPreview = ({
  className,
  children,
  classes,
  disableOverlay = false,
  onUnload,
  onClick,
  ...others
}: HvFileUploaderPreviewProps) => {
  useEffect(() => {
    return () => {
      onUnload?.();
    };
  }, [onUnload]);

  if (onClick) {
    return (
      <StyledButton
        icon
        className={clsx(
          className,
          classes?.previewButton,
          fileUploaderPreviewClasses.previewButton
        )}
        onClick={onClick}
        variant="secondaryGhost"
        {...others}
      >
        {children}
        {!disableOverlay && (
          <>
            <StyledOverlay
              className={clsx(
                classes?.overlay,
                fileUploaderPreviewClasses.overlay
              )}
              aria-hidden="true"
            />
            <StyledPreviewIcon />
          </>
        )}
      </StyledButton>
    );
  }

  return children;
};
