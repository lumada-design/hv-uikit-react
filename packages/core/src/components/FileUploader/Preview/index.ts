export { default as fileUploaderPreviewClasses } from "./previewClasses";
export * from "./previewClasses";
export * from "./Preview";
