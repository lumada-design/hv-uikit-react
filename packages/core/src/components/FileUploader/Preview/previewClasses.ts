import { getClasses } from "utils";

export type HvFileUploaderPreviewClasses = {
  previewButton?: string;
  overlay?: string;
};

const classKeys: string[] = ["previewButton", "overlay"];

const fileUploaderPreviewClasses = getClasses<HvFileUploaderPreviewClasses>(
  classKeys,
  "HvFileUploaderPreview"
);

export default fileUploaderPreviewClasses;
