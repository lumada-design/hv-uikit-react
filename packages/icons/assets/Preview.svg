<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path fill="#414141" d="M11 8a3 3 0 1 1-3-3 3 3 0 0 1 3 3m5 0s-3.58 6-8 6-8-6-8-6 4-6 8-6 8 6 8 6m-1.2.03a22 22 0 0 0-2-2.32C11.01 3.94 9.35 3 8 3s-3 .93-4.77 2.68a22 22 0 0 0-2.02 2.35 18 18 0 0 0 1.85 2.28C4.25 11.53 6.08 13 8 13s3.73-1.45 4.91-2.67a18 18 0 0 0 1.88-2.3z"/></svg>