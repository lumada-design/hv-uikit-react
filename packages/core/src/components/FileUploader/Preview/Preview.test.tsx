import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { HvFileUploaderPreview } from "./Preview";

describe("FileUploaderPreview", () => {
  it("should be defined", () => {
    const { container } = render(
      <HvFileUploaderPreview>
        <div>Hello</div>
      </HvFileUploaderPreview>
    );

    expect(container).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(
      <HvFileUploaderPreview>
        <div>Hello</div>
      </HvFileUploaderPreview>
    );

    expect(container).toMatchSnapshot();
  });

  it("should just render its children", () => {
    const { container, queryByRole, getByTestId } = render(
      <HvFileUploaderPreview>
        <div data-testid="child">Hello</div>
      </HvFileUploaderPreview>
    );

    const button = queryByRole("button");

    expect(button).not.toBeInTheDocument();

    const child = getByTestId("child");

    expect(child).toBeInTheDocument();
    expect(container.firstChild).toBe(child);
  });

  it("should render its children inside a button if onClick is provided", () => {
    const onClickMock = vi.fn();

    const { container, getByRole, getByTestId } = render(
      <HvFileUploaderPreview onClick={onClickMock}>
        <div data-testid="child">Hello</div>
      </HvFileUploaderPreview>
    );

    const button = getByRole("button");

    expect(button).toBeInTheDocument();

    const child = getByTestId("child");

    expect(child).toBeInTheDocument();
    expect(container.firstChild).toBe(button);
  });

  it("should call onUnload when unmounted", () => {
    const onUnloadMock = vi.fn();

    const { unmount } = render(
      <HvFileUploaderPreview onUnload={onUnloadMock}>
        <div data-testid="child">Hello</div>
      </HvFileUploaderPreview>
    );

    expect(onUnloadMock).toHaveBeenCalledTimes(0);

    unmount();

    expect(onUnloadMock).toHaveBeenCalledTimes(1);
  });

  it("should call onClick when clicked", async () => {
    const onClickMock = vi.fn();

    const { getByRole } = render(
      <HvFileUploaderPreview onClick={onClickMock}>
        <div data-testid="child">Hello</div>
      </HvFileUploaderPreview>
    );

    const button = getByRole("button");

    expect(onClickMock).toHaveBeenCalledTimes(0);

    await userEvent.click(button);

    expect(onClickMock).toHaveBeenCalledTimes(1);
  });
});
