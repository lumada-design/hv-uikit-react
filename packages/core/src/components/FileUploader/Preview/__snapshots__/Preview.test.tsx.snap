// Vitest Snapshot v1

exports[`FileUploaderPreview > should render correctly 1`] = `
<div>
  <div>
    Hello
  </div>
</div>
`;
